### PR TITLE
Adding memguardian + various optimizations

### PR DESCRIPTION
**Pull request by Mzack9999**
## Proposed changes
- Introducing the `memguardian` component that actively monitor used memory. Activation is optional via `export MEMGUARDIAN=1`. If the used RAM is above the threshold (default `75%` of available RAM). It reduces parallel threading of templates to either `MEMGUARDIAN_THREADS` if defined, or 1/5 of the total original threading. Also if `MEMGUARDIAN_ALLOC` is defined, it caps the maximum amount of buffers that can be allocated in the pool. When memory conditions are back to normal, the slow down mechanisms are automatically reverted.
- collecting events within ScanContext is now optional, as it's not used anywhere and it was preventing `gc` to recycle large maps of data (and probably a co-cause of #4836) 

## Checklist
- [ ] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Example
Closes https://github.com/projectdiscovery/nuclei/issues/4756

```console
$ MEMGUARDIAN=1 go run . -l 790targets.txt -jsc 110 -stats -tags fuzz,fuzzing,brute-force,bruteforce,brute,dos,local,code  -severity critical -rl 0
...
[INF] Templates loaded for current scan: 6
[INF] Executing 4 signed templates from projectdiscovery/nuclei-templates
[WRN] Executing 3 unsigned templates. Use with caution.
[INF] Targets loaded for current scan: 790
[0:00:06] | Templates: 6 | Hosts: 790 | RPS: 1796 | Matched: 0 | Errors: 0 | Requests: 11227/6737120 (0%)
...
[0:35:50] | Templates: 6 | Hosts: 790 | RPS: 686 | Matched: 0 | Errors: 37507 | Requests: 6417135/6737120 (95%)
```

### CODE REVIEW

File: pkg/protocols/http/http.go [hunk 1]
```diff
@@ -388,8 +390,13 @@ func (request *Request) Compile(options *protocols.ExecutorOptions) error {
 		}
 	}
 	if len(request.Payloads) > 0 {
+		// specifically for http requests high concurrency and and threads will lead to memory exausthion, hence reduce the maximum parallelism
+		if protocolstate.MemGuardian.Warning.Load() {
+			request.Threads = 5
```
Comment (stan-threatmate) [inline pkg/protocols/http/http.go:395]: can this be configurable as a setting?

File: pkg/protocols/common/protocolstate/state.go [hunk 2]
```diff
@@ -15,16 +16,34 @@ import (
...
 
 // Init creates the Dialer instance based on user configuration
 func Init(options *types.Options) error {
 	if Dialer != nil {
 		return nil
 	}
+	if MemGuardian == nil {
+		mg, err := memguardian.New(
+			memguardian.WitInterval(time.Second),
+			memguardian.WithRatioWarning(75),
```
Comment (stan-threatmate) [inline pkg/protocols/common/protocolstate/state.go:36]: can the ratio be configurable as a setting?

File: pkg/protocols/http/http.go (hunk 1 above)
Comment (Mzack9999) [reply pkg/protocols/http/http.go:395]: Definitely - I'm going to refactor it and allow configuration via environment variable

File: pkg/protocols/common/protocolstate/state.go (hunk 2 above)
Comment (Mzack9999) [reply pkg/protocols/common/protocolstate/state.go:36]: same as my previous comment

**Comment by Mzack9999** [comment]
Optional behavior - Merging

**Pull request merged**